File: Plugins/Orimath.Basics.View/ViewModels/NewPaperDialogViewModel.cs
using System.Windows.Input;
using Orimath.Controls;
using Orimath.Plugins;
using ApplicativeProperty;

namespace Orimath.Basics.View.ViewModels
{
    public class NewPaperDialogViewModel : NotifyPropertyChanged
    {
        private readonly IMessenger _messenger;
        private readonly IDispatcher _dispatcher;
        private readonly NewPaperExecutor _executor;

        public NewPaperDialogViewModel(IMessenger messenger, IDispatcher dispatcher, NewPaperExecutor executor)
        {
            _messenger = messenger;
            _dispatcher = dispatcher;
            _executor = executor;
            ExecuteCommand = Prop.True.ToCommand(_ => Execute());
            CloseCommand = messenger.CloseDialogCommand;

            if (_executor.NewPaperType.IsSquare)
            {
                _isSquareSelected = true;
            }
            else if (_executor.NewPaperType is NewPaperType.Rectangle rectangle)
            {
                _isRectangleSelected = true;
                _width = rectangle.width;
                _height = rectangle.height;
            }
            else if (_executor.NewPaperType is NewPaperType.RegularPolygon polygon)
            {
                _isPolygonSelected = true;
                _numberOfPolygon = polygon.number;
            }
        }

        private bool _isSquareSelected;
        public bool IsSquareSelected
        {
            get => _isSquareSelected;
            set
            {
                if (SetValue(ref _isSquareSelected, value) && value)
                    SetSquare();
            }
        }

        private bool _isRectangleSelected;
        public bool IsRectangleSelected
        {
            get => _isRectangleSelected;
            set
            {
                if (SetValue(ref _isRectangleSelected, value) && value)
                    SetRectangle();
            }
        }

        private bool _isPolygonSelected;
        public bool IsPolygonSelected
        {
            get => _isPolygonSelected;
            set
            {
                if (SetValue(ref _isPolygonSelected, value) && value)
                    SetPolygon();
            }
        }

        private double _width = 1.0;
        public double Width
        { 
            get => _width;
            set
            {
                if (SetValue(ref _width, value) && IsRectangleSelected)
                    SetRectangle();
            }
        }

        private double _height = 1.0;
        public double Height
        {
            get => _height;
            set
            {
                if (SetValue(ref _height, value) && IsRectangleSelected)
                    SetRectangle();
            }
        }

        private int _numberOfPolygon = 3;
        public int NumberOfPolygon
        { 
            get => _numberOfPolygon;
            set
            {
                if (SetValue(ref _numberOfPolygon, value) && IsPolygonSelected)
                    SetPolygon();
            }
        }

        private void SetSquare()
        {
            _executor.NewPaperType = NewPaperType.Square;
        }

        private void SetRectangle()
        {
            _executor.NewPaperType = NewPaperType.NewRectangle(Width, Height);
        }

        private void SetPolygon()
        {
            _executor.NewPaperType = NewPaperType.NewRegularPolygon(NumberOfPolygon);
        }

        public async void Execute()
        {
            await _dispatcher.OnBackgroundAsync(_executor.NewPaper);
            _messenger.CloseDialog();
        }

        public ICommand ExecuteCommand { get; }

        public ICommand CloseCommand { get; }
    }
}

File: Plugins/Orimath.Basics.View/ViewModels/NewPaperDialogViewModel.fs
﻿namespace Orimath.Basics.View.ViewModels
open Orimath.Basics
open Orimath.Controls
open Orimath.Plugins
open ApplicativeProperty
open ApplicativeProperty.PropOperators

type NewPaperDialogViewModel(messenger: IMessenger, dispatcher: IDispatcher, executor: NewPaperExecutor) =
    inherit NotifyPropertyChanged()
    let isSquareSelected = Prop.value false
    let isRectangleSelected = Prop.value false
    let isPolygonSelected = Prop.value false
    let width = Prop.value 1.0
    let height = Prop.value 1.0
    let numberOfPolygon = Prop.value 3
    do
        match executor.NewPaperType with
        | Square ->
            isSquareSelected .<- true
        | Rectangle(w, h) ->
            isRectangleSelected .<- true
            width .<- w
            height .<- h
        | RegularPolygon(number) ->
            isPolygonSelected .<- true
            numberOfPolygon .<- number
    do isSquareSelected.Add(fun b -> if b then executor.NewPaperType <- Square)
    do (isRectangleSelected, width, height)
       |||> Prop.map3 (fun b w h -> b, Rectangle(w, h))
       |> Observable.add (fun (b, pt) -> if b then executor.NewPaperType <- pt)
    do (isPolygonSelected, numberOfPolygon)
       ||> Prop.map2(fun b n -> b, RegularPolygon(n))
       |> Observable.add (fun (b, pt) -> if b then executor.NewPaperType <- pt)

    member val SquareText = messenger.LocalizeWord("{basic/Dialog.NewPaper.Square}Square")
    member val RectangleText = messenger.LocalizeWord("{basic/Dialog.NewPaper.Rectangle}Rectangle")
    member val WidthText = messenger.LocalizeWord("{basic/Dialog.NewPaper.Width}Width: ")
    member val HeightText = messenger.LocalizeWord("{basic/Dialog.NewPaper.Height}Height: ")
    member val PolygonText = messenger.LocalizeWord("{basic/Dialog.NewPaper.Polygon}Polygon")
    member val RegularText = messenger.LocalizeWord("{basic/Dialog.NewPaper.Regular}Regular")
    member val GonText = messenger.LocalizeWord("{basic/Dialog.NewPaper.Gon}-gon")

    member _.IsSquareSelected = isSquareSelected
    member _.IsRectangleSelected = isRectangleSelected
    member _.IsPolygonSelected = isPolygonSelected
    member _.Width = width |> Prop.mapBoth id (fun w -> if w <= 0.0 then 1.0 else w)
    member _.Height = height |> Prop.mapBoth id (fun h -> if h <= 0.0 then 1.0 else h)
    member _.NumberOfPolygon = numberOfPolygon

    member val ExecuteCommand = Prop.ctrue |> Prop.command (fun _ ->
        dispatcher.Background {
            executor.NewPaper()
            dispatcher.UI.Invoke(messenger.CloseDialog)
        })

    member _.CloseCommand = messenger.CloseDialogCommand